src/main/java/com/google/devtools/build/lib/skyframe/PrerequisiteProducer.java

```diff
@@ -318,6 +318,30 @@ public boolean evaluate(
     //  would exit this SkyFunction and restart it when permits were available.
     semaphoreLocker.acquireSemaphore();
     try {
+      // Check target compatibility before requesting toolchains - known missing toolchains are a
+      // valid reason for declaring a target incompatible and should not result in the target
+      // failing to build rather than being skipped as incompatible.
+      // Non-rule targets and those that are part of the toolchain resolution system do not support
+      // target compatibility checking anyway.
+      if (targetAndConfiguration.getTarget() instanceof Rule
+          && ((Rule) targetAndConfiguration.getTarget()).useToolchainResolution()) {
+        platformInfo = loadTargetPlatformInfo(env);
+        // loadTargetPlatformInfo may return null even when no deps are missing.
+        if (env.valuesMissing()) {
+          return false;
+        }
+
+        configConditions = computeConfigConditions(env, targetAndConfiguration, transitivePackages,
+            platformInfo, transitiveRootCauses);
+        if (configConditions == null) {
+          return false;
+        }
+
+        if (!checkForIncompatibleTarget(env, state, transitivePackages)) {
```

Comment:
Minor nitpick: but would it help to make this a static method?

I didn't understand why configConditions is moved to line 334 until I realized checkForIncompatibleTarget references it. Maybe it'd be easier to follow if we just passed all state as parameters?

Reply:
Fully agree, this also confused me at first. Changed.